test/test_gen_backend_stubs.py

```diff
@@ -0,0 +1,100 @@
+import os
+import tempfile
+
+from torch.testing._internal.common_utils import TestCase, run_tests
+import tools.codegen.gen_backend_stubs
+
+path = os.path.dirname(os.path.realpath(__file__))
+gen_backend_stubs_path = os.path.join(path, '../tools/codegen/gen_backend_stubs.py')
+
+# gen_backend_stubs.py is an integration point that is called directly by external backends.
+# The tests here are to confirm that badly formed inputs result in reasonable error messages.
+class TestGenBackendStubs(TestCase):
+
+    def get_errors_from_gen_backend_stubs(self, yaml_str: str) -> str:
+        with tempfile.NamedTemporaryFile(mode='w') as fp:
+            fp.write(yaml_str)
+            fp.flush()
+            try:
+                tools.codegen.gen_backend_stubs.run(fp.name, '', True)
+            except AssertionError as e:
+                # Scrub out the temp file name from any error messages to simplify assertions.
+                return str(e).replace(fp.name, '')
+            self.fail('Expected gen_backend_stubs to raise an AssertionError, but it did not.')
+
+    def test_missing_backend(self):
+        yaml_str = '''\
+cpp_namespace: torch_xla
+supported:
+- abs'''
+        output_error = self.get_errors_from_gen_backend_stubs(yaml_str)
+        self.assertExpectedInline(output_error, '''You must provide a value for "backend"''')
+
+    def test_empty_backend(self):
+        yaml_str = '''\
+backend:
+cpp_namespace: torch_xla
+supported:
+- abs'''
+        output_error = self.get_errors_from_gen_backend_stubs(yaml_str)
+        self.assertExpectedInline(output_error, '''You must provide a value for "backend"''')
+
+    def test_missing_cpp_namespace(self):
+        yaml_str = '''\
+backend: XLA
+supported:
+- abs'''
+        output_error = self.get_errors_from_gen_backend_stubs(yaml_str)
+        self.assertExpectedInline(output_error, '''You must provide a value for "cpp_namespace"''')
+
+    def test_whitespace_cpp_namespace(self):
+        yaml_str = '''\
+backend: XLA
+cpp_namespace:\t
+supported:
+- abs'''
+        output_error = self.get_errors_from_gen_backend_stubs(yaml_str)
+        self.assertExpectedInline(output_error, '''You must provide a value for "cpp_namespace"''')
+
+    def test_empty_supported(self):
+        yaml_str = '''\
+backend: XLA
+cpp_namespace: torch_xla
+supported:
+'''
+        output_error = self.get_errors_from_gen_backend_stubs(yaml_str)
+        self.assertExpectedInline(output_error, '''expected "supported" to be a list, but got: None (of type <class 'NoneType'>)''')
```

Comment:
Arguably this should work lol

Reply:
yeah that's fair (I'll just get that case working before I land). I assume the use case you're picturing a new backend that wants to set up all of the boilerplate required for integration with us, without being required to write any kernels first?


Separately from that is the singleton case that I have a test for: `supported: __`. I tentatively think it's fine to leave this as an error, since (a) it doesn't seem like a useful thing to support (more "flexible" yaml?), and (b) If you do something dumb like `supported: XLA`, you'll get a worse error.